From: David Howells <dhowells@redhat.com>
Date: Mon, 19 Oct 2015 11:20:28 +0100
Subject: KEYS: Don't permit request_key() to construct a new keyring
Origin: https://git.kernel.org/linus/911b79cde95c7da0ec02f48105358a36636b7a71

If request_key() is used to find a keyring, only do the search part - don't
do the construction part if the keyring was not found by the search.  We
don't really want keyrings in the negative instantiated state since the
rejected/negative instantiation error value in the payload is unioned with
keyring metadata.

Now the kernel gives an error:

	request_key("keyring", "#selinux,bdekeyring", "keyring", KEY_SPEC_USER_SESSION_KEYRING) = -1 EPERM (Operation not permitted)

Signed-off-by: David Howells <dhowells@redhat.com>
[carnil: Backported for 3.2: adjust context. Search context structure only
introduced in 4bdf0bc300314141e5475e145acb8b5ad846f00d]
---
 security/keys/request_key.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/security/keys/request_key.c
+++ b/security/keys/request_key.c
@@ -472,6 +472,9 @@ static struct key *construct_key_and_lin
 
 	kenter("");
 
+	if (type == &key_type_keyring)
+		return ERR_PTR(-EPERM);
+
 	ret = construct_get_dest_keyring(&dest_keyring);
 	if (ret)
 		goto error;
